src/factor_graph.h

```diff
@@ -33,7 +33,10 @@ class FactorGraphDescriptor {
   /** number of edges */
   num_edges_t num_edges;
 
-  /** number of all weights */
+  /**
+   * number of weights. this is still kept here to maintain format of meta,
+   * and is also useful when loading weights
+   */
   num_weights_t num_weights;
```

Comment:
Ideally, these shall get decoupled and InferenceResult (or LearningState) will be constructed independently from others (e.g., FactorGraph or GibbsSampler).

Not sure how much overhead vector has over plain array access, but I think we shouldn't give up anything since weight access is on a really hot path.